File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/MassStorageSample/src/com/hp/workpath/sample/massstoragesample/task/GetStorageListTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.massstoragesample.task;

import android.os.Handler;
import android.os.Looper;
import android.view.View;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.massstorage.MassStorageInfo;
import com.hp.workpath.api.massstorage.MassStorageService;
import com.hp.workpath.sample.massstoragesample.Logger;
import com.hp.workpath.sample.massstoragesample.MainActivity;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class GetStorageListTask {

    private final WeakReference<MainActivity> mContextRef;
    private Result mResult;

    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public GetStorageListTask(MainActivity context) {
        this.mContextRef = new WeakReference<>(context);
        this.mResult = new Result();
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        onPostExecute(MassStorageService.getStorageList(mContextRef.get(), mResult));
                    } catch (Throwable t) {
                        Logger.showResult(null, "MassStorageService.getStorageList " + t.getMessage());
                        executor.shutdown();
                        onPostExecute(null);
                    }
                }
            });
        } catch (Exception e) {
            Logger.showResult(null, "MassStorageService.getStorageList " + e.getMessage());
            onPostExecute(null);
            executor.shutdown();
        }
    }

    private void onPostExecute(List<MassStorageInfo> massStorageInfoList) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                mContextRef.get().showProgress(View.GONE);
                Logger.showResult(mContextRef.get(), "MassStorageService.getStorageList", mResult);
                if (massStorageInfoList != null && mResult.getCode() == Result.RESULT_OK) {
                    mContextRef.get().loadStorageList(massStorageInfoList);
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/MassStorageSample/src/com/hp/workpath/sample/massstoragesample/task/GetStorageListTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.massstoragesample.task

import android.view.View
import com.hp.workpath.api.Result
import com.hp.workpath.api.massstorage.MassStorageInfo
import com.hp.workpath.api.massstorage.MassStorageService
import com.hp.workpath.sample.massstoragesample.Logger
import com.hp.workpath.sample.massstoragesample.MainActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

class GetStorageListTask(context: MainActivity) {
    private val mContextRef: WeakReference<MainActivity> = WeakReference(context)
    private var mThrowable: Throwable? = null
    private val mResult: Result = Result()

    suspend fun execute() {
        var massStorageInfoList: MutableList<MassStorageInfo>? = null
        try {
            mContextRef.get()?.run {
                massStorageInfoList = MassStorageService.getStorageList(this, mResult)
            }
        } catch (t: Throwable) {
            mThrowable = t
        }
        onPostExecute(massStorageInfoList)
    }

    private suspend fun onPostExecute(massStorageInfoList: MutableList<MassStorageInfo>?) {
        withContext(Dispatchers.Main) {
            mContextRef.get()?.run {
                showProgress(View.GONE)
                Logger.showResult(this, "MassStorageService.getStorageList", mResult)
                if (massStorageInfoList != null && mResult.code == Result.RESULT_OK) {
                    loadStorageList(massStorageInfoList)
                } else if (mThrowable != null) {
                    Logger.showResult(this, "MassStorageService.getStorageList ${mThrowable?.message}")
                }
            }
        }
    }
}